lib/xdrgen/generators/go.rb

```diff
@@ -549,6 +531,259 @@ def render_encode_to_body(out, var, type, self_encode:)
         out.puts "  }"
       end
 
+      def render_struct_decode_from_interface(out, struct)
+        name = name(struct)
+        out.puts "// DecodeFrom decodes this value using the Decoder."
+        out.puts "func (s *#{name}) DecodeFrom(d *xdr.Decoder) (int, error) {"
+        out.puts "  var err error"
+        out.puts "  var n, nTmp int"
+        declared_variables = []
+        struct.members.each do |m|
+          mn = name(m)
+          render_decode_from_body(out, "s.#{mn}", m.type, declared_variables: declared_variables, self_encode: false)
+        end
+        out.puts "  return n, nil"
+        out.puts "}"
+        out.break
+      end
+
+      def render_union_decode_from_interface(out, union)
+        name = name(union)
+        out.puts "// DecodeFrom decodes this value using the Decoder."
+        out.puts "func (u *#{name}) DecodeFrom(d *xdr.Decoder) (int, error) {"
+        out.puts "  var err error"
+        out.puts "  var n, nTmp int"
+        render_decode_from_body(out, "u.#{name(union.discriminant)}", union.discriminant.type, declared_variables: [], self_encode: false)
+        switch_for(out, union, "u.#{name(union.discriminant)}") do |arm, kase|
+          out2 = StringIO.new
+          if arm.void?
+            out2.puts "// Void"
+          else
+            mn = name(arm)
+            type = arm.type
+            out2.puts "  u.#{mn} = new(#{reference arm.type})"
+            render_decode_from_body(out2, "(*u.#{mn})",type, declared_variables: [], self_encode: false)
+          end
+          out2.puts "  return n, nil"
+          out2.string
+        end
+        unless union.default_arm.present?
+          out.puts "  return n, fmt.Errorf(\"#{name(union.discriminant)} (#{reference union.discriminant.type}) switch value '%d' is not valid for union #{name}\", u.#{name(union.discriminant)})"
+        end
+        out.puts "}"
+        out.break
+      end
+
+      def render_enum_decode_from_interface(out, typedef)
+        name = name(typedef)
+        type = typedef
+        out.puts <<-EOS.strip_heredoc
+        // DecodeFrom decodes this value using the Decoder.
+        func (e *#{name}) DecodeFrom(d *xdr.Decoder) (int, error) {
+          v, n, err := d.DecodeInt()
+          if err != nil {
+            return n, err
+          }
+          if _, ok := #{private_name type}Map[v]; !ok {
```

Comment:
💭 Given that most of our enums start at zero and are sequential, and given that use of the map will now be in the hot path, I wonder if for enums that we detect have that property, if we could simply replace this with a `if v >= len(...Map) { // error }`? We don't have to do this, and we definitely don't have to do this in this PR, but maybe something to evaluate if there is a saving here over hitting the map.

Reply:
Good idea, However:

There are many enums with negative values. Some examples are: `BucketEntryType`, `CreateAccountResultCode`, `PaymentResultCode`, `PathPaymentStrictReceiveResultCode` ....

Not even a range is generally a possibility since the XDR definition can contain scattered values.

We could infer a range where available and use that, but I think it is an overoptimization (without further proof of a bottleneck) and definitely out of the scope of this PR.

Let's later see if further profiling reveals map access as a problem.